### DISCUSSION TITLE
Support for dynamic VEX retrieval from "External References" of SBOMs

### DISCUSSION THREAD

**Discussion by RingoDev** [aquasecurity/trivy#8047] [Ideas]
## What is this about?

I am wondering if it makes sense for the most feature complete OS SBOM scanner (namely Trivy) to support the dynamic loading of VEX statements specified via an "external reference" in CycloneDX or SPDX SBOMs.

## What am I talking about
Both CycloneDx and SPDX SBOM standards have the possibility to included "externalReferences" with a strict enum of possibly types of external references / documents that are related to that SBOM. What is especially interesting for me here is the ["exploitability-statement" type of CycloneDx](https://github.com/CycloneDX/specification/blob/268138039498a4f390ddc5d7e7e78857a951986f/schema/bom-1.6.schema.json#L1763C14-L1763C38) and the ["vulnerabilityExploitabilityAssessment" type of SPDX](https://github.com/spdx/spdx-3-model/blob/14ba670c930a06680478bfdca9616ebd7f8b84ac/model/Core/Vocabularies/ExternalRefType.md?plain=1#L64).

As this field is explicitly allowed to contain URLs I believe it would be a valid usecase to make a lookup to such an URL and check if the resource located at that URL is a supported VEX format that can be applied to filter the produced output.

### What would be the value?
This could be used to have a very "efficient" way of retrieving VEX information specifically related to the SBOM that is being scanned. Compared to the "vex-repo" approach, in the best case this would only require to retrieve the VEX information that is explicitly relevant to the scanned SBOM. This would even open up possibilities for publishers/vendors to generate VEX files dynamically on the fly and distribute them via an API.

Furthermore, this gives vendors that are already publishing SBOMs (e.g. as part of their container images with Cosign and In-Toto attestations) a way for these SBOMs to point at where the dynamic VEX information for that SBOM can be retrieved.

### What could be downsides?

* Even though there is a specific type for "exploitability-statement" this could point to any kind of external reference and does not even require to be a valid URL.
* I am not sure how many vendors are/will be reyling on that - except that I would like to make use of it 😅.
* This would be dynamically trying to retrieve information from a scanned SBOM and applying filtering possibly without the initiator of the scan being aware of it. It might make sense to have explicity logging if such an external reference is resolved and used, or even hide the feature behind a flag such as "--vex-resolve-external-references".
* This is yet another way of where VEX information might be coming from, whereas I see that you are currenly already working on speccing out thte Vex-Hub and Vex-Repository functionality - which possibly increases maintenance

### How much work do I think that is?

Actually with Trivy already supporting VEX filtering and the ability to handle different VEX formats this is "only" another way of specifying where the VEX information is coming from that should be used to "filter" the results of an SBOM scan.

### Who would need to do the work

I am currently pretty deep in that topic as I am working on a master's thesis in that field and would be willing to contribute if that seems to be a reasonable feature to have.

### Target

SBOM

### Scanner

Vulnerability

**Comment by knqyf263** [comment]
I think it should be discussed with the CycloneDX and SPDX teams first. Once it's standardized, we're happy to add support for that. Otherwise, nobody will use the field for this purpose, and implementing it in Trivy will be meaningless.

Also, the VEX Repository Specification potentially allows the indexing of external VEX documents. 
https://github.com/aquasecurity/vex-repo-spec/issues/2

This can achieve this benefit you pointed out.

> in the best case this would only require to retrieve the VEX information that is explicitly relevant to the scanned SBOM.

**Reply by RingoDev** [reply]
I will take it up with the CDX folks first, thanks for the feedback.

I think it would solve a bit of a different usecase than I believe the VEX Repo spec is coming from. With tightly coupling a scanned SBOM to the VEX information that is only relevant to be applied for that SBOM just seems like a different approach than scanning through a potentially large list of VEX files and matching based on PURLS. 

Especially if you have have a lot of different container images on all of the main public container registries. It seems like a pain to maintain all of these different oci PURLs to point at the correct VEX repo. Whereas the SBOMs attested on those containers already could easily propagate where the VEX source for that container can be found.

Also I have not found information about your standardization efforts or the governance structure of the VEX-Repo spec. I would be more eager to try it out if I know what the future of that spec is shaping to.

Nevertheless, thanks for trying to push the whole distribution fo VEX topic forward, I think that is very much needed.

**Reply by knqyf263** [reply]
>Also I have not found information about your standardization efforts or the governance structure of the VEX-Repo spec. I would be more eager to try it out if I know what the future of that spec is shaping to.

We're discussing here.
https://docs.google.com/document/d/1Db-Izh2PWjIXH7mS54K1cHWsttq4_bxg9v3_YI2enyU/edit?tab=t.0

We haven't had a chance to discuss it since recent meetings were canceled, but we hope to move forward in a few months.

**Reply by knqyf263** [reply]
>Especially if you have have a lot of different container images on all of the main public container registries. It seems like a pain to maintain all of these different oci PURLs to point at the correct VEX repo. Whereas the SBOMs attested on those containers already could easily propagate where the VEX source for that container can be found.

You are correct, but the placement of the SBOM is not standardized either. For container images, [the OCI referrer](https://trivy.dev/latest/docs/target/container_image/#discovery) may be available, but it is a long way to go, as you need to start with the step of ensuring that you get the SBOM first. We have chosen the shortest practical method, even if it is not ideal. So, it can be replaced by other methods in the future.

**Reply by RingoDev** [reply]
Ah interesting, so I guess at the moment you are not able to discover SBOMs attested on the OCI registry via cosign automatically (yet)?

EDIT: Just found the docs that show that cosign is required to be run before trivy can scan the retrieved SBOM: https://trivy.dev/latest/docs/supply-chain/attestation/sbom/#scanning

This would sound like a nice addition if you were able to discover such SBOMs in a single call.

**Reply by knqyf263** [reply]
>Ah interesting, so I guess at the moment you are not able to discover SBOMs attested on the OCI registry via cosign automatically (yet)?

No, I mean there are many ways to attach SBOMs in OCI registries. Cosign does that using their approach, which is not standardized in OCI. Also, it's anyway applicable only to OCI registries. How about other artifacts or git repositories? AFAIK, there is no standard for SBOM discovery.

**Reply by RingoDev** [reply]
You are absolutely correct here. Also due to the fact that ecosystems differ so much and distribution channels are very different for something e.g. packaged as a container vs a CLI tool or a binary downloadable e.g. from a GitHub release page.

This makes a strong use case for the work being done on the [Transparency Exchange API](https://github.com/CycloneDX/transparency-exchange-api) where you have a strongly specified API that serves/points to artifacts such as SBOM or VEX related to a Transparency Exchange Identifier (TEI). 

How to propagate such an identifier with the artifact then, might just have the same discovery problem as you currently have though, that I do not think has had enough attention at this point.

**Comment by RingoDev** [comment]
To reiterate on this after the christmas break. 

It seems as if there are already "some" examples of package maintainers using these external references to allow the discovery of dynamic documents (in this case VDR and not VEX) from their SBOMs: https://www.sonatype.com/blog/sbom-vdr-and-maven-transforming-the-apache-logging-experience-to-a-common-pattern

Also just to drive the point across, I believe this has already been "standardized" as these are the definitions that CycloneDx and SPDX use to describe the field that seem very self-explanatory to me:

* SPDX: _A reference to a Vulnerability Exploitability eXchange (VEX) statement which provides information on whether a product is impacted by a specific vulnerability in an included package and, if affected, whether there are actions recommended to remediate. See also [NTIA VEX one-page summary](https://ntia.gov/files/ntia/publications/vex_one-page_summary.pdf)._
* CycloneDx: _A Vulnerability Exploitability eXchange (VEX) which asserts the known vulnerabilities that do not affect a product, product family, or organization, and optionally the ones that do. The VEX should include the analysis and findings describing the impact (or lack of impact) that the reported vulnerability has on the product, product family, or organization.",_

Whether how widely it is/would be adopted is a different question ...

To me it would be a more "simplistic" and tool agnostic approach to dynamically discover a VEX document than the VEX-Hub could currently provide. It could even be used alongside the structure of the VEX-Hub to point to a file stored in GitHub.


I have a branch on trivy ready that shows how it would fit in with what is already there in terms of SBOM and VEX handling for my own POC. I could raise it as a PR if there is interest.


EDIT: I stumbled over an issue on Grype where it seems like @sethmlarson has outlined the exact same proposal with a nice diagram: https://github.com/anchore/grype/issues/1365#issuecomment-1823538226

**Reply by knqyf263** [reply]
It sounds reasonable. The next step is to decide how we enable this feature. Since it's possible to call many HTTP requests, we should disable it by default (VEX Hub is also not enabled by default). Do you have any ideas?

**Reply by RingoDev** [reply]
I would put it in line with the `--vex repo` or `--vex oci` flags and add something like `--vex externalRef`.  Does that sound reasonable?

**Reply by knqyf263** [reply]
I'd include "SBOM". Otherwise, what `externalRef` refers to is unclear.

- `--vex sbom`
- `--vex sbom-ref`
- `--vex sbom-external-ref`

**Reply by RingoDev** [reply]
Makes sense to me

Of these options I believe `--vex sbom` could be misleading in the way that SBOMs can also come with integrated vulnerability information in the SBOM itself.

I'd vote for `--vex sbom-ref` as being most suitable

**Reply by knqyf263** [reply]
OK, it sounds reasonable. It would be experimental anyway, like other VEX options. We can see the feedback from the community and rename it later if needed.

**Comment by RingoDev** [comment]
I opened #8254 that would be adding initial support of this for cyclonedx SBOMs. Review at your convenience

**Comment by macedogm** [comment]
Regarding:

> I am not sure how many vendors are/will be reyling on that - except that I would like to make use of it 😅.

And:

> Otherwise, nobody will use the field for this purpose, and implementing it in Trivy will be meaningless.

You can certainly count SUSE Rancher on that! We are currently **investigating, still on an early phase,** the viability of providing a minimal SBOM (following NTIA's ['The Minimum Elements For a Software Bill of Materials](https://www.ntia.gov/sites/default/files/publications/sbom_minimum_elements_report_0.pdf)') inside of our images and with our own curated CVE database, similar to what Bitnami does.

The reason is to provide more accurate information about the applicability of CVEs in our images. Having the possibility of using a field to specify a VEX report, even if pointing directly to our "bigger and unified" [VEX report](https://github.com/rancher/vexhub/blob/main/reports/rancher.openvex.json), for example, would be a big win IMO.

That can make VEX fully transparent to users that are not interested in having to configure it or that might distrust VEX Hubs (be them from Aqua/Trivy, even Rancher's etc.). I do see some users that still only trust what comes directly from the vendor's CVE database and nothing else, even if provided directly by the vendor (the education part is an on going effort).

**Comment by RingoDev** [comment]
Feature is included for CycloneDX SBOMs with #8254 in latest v0.60.0 release of Trivy (#8327). Thanks for the support and clean-up @knqyf263 🥳
